### ISSUE TITLE
pytorch2caffe failed at add operation 

### ISSUE THREAD

**Issue by andyhahaha** [hahnyuan/nn_tools#13]
I'm trying convert RefineDet to caffe model. RefineDet only contain regular CNN module(conv2d, convtranspose2d, maxpool, batchnorm).
https://github.com/luuuyi/RefineDet.PyTorch
However, pytorch2caffe failed at add operation as follow:
![image](https://user-images.githubusercontent.com/14272969/54815593-58015600-4ccd-11e9-9a50-40d56328a9c6.png)

This add operation is from here
https://github.com/luuuyi/RefineDet.PyTorch/blob/0e4b24ce07245fcb8c48292326a731729cc5746a/layers/modules/l2norm.py#L20
Why pytorch2caffe failed at a simple operation?
Thanks for help!

**Comment by hahnyuan** [comment]
Hi, andyhahaha

Thank you for your remind.
The question is that the Reduction Layer in Caffe only reduction along ALL "tail" axes is supported. See http://caffe.berkeleyvision.org/tutorial/layers/reduction.html. So there is no directly way to transfer the sum(dim=1, keepdim=True)

There is one way to solve this problem. Replacing the forward class of LRN class with a unsupported operation such as F.relu6(), that this layer will transfered to a Python Layer. So you can change the type of these layer to LRN in transferred caffe prototxt. 

Wish this will help you.

Zhihang

**Comment by andyhahaha** [comment]
Hi, Zhihang

Thanks for reply!
But I think this code haven't run to the sum(dim=1, keepdim=True) yet.
I set the pdb and found that it stop at the function "_add", between x.pow(2).sum(dim=1, keepdim=True).sqrt() and self.eps.
Do you know why it would failed at add function?

Thanks!

**Comment by hahnyuan** [comment]
Oh. That is in ___add__ function, the float adding was not considered. Only the Element wise adding is considered. I have solved this problem at the latest master branch.

**Comment by andyhahaha** [comment]
Thanks a lot!
It works.

**Comment by andyhahaha** [comment]
It works! Thanks.
But after that I encounter another issue.
This error message said there is operation not implemented.
Is it means cat or view?  But I can found these two function implemented in your code. 
![image](https://user-images.githubusercontent.com/14272969/54904136-96d71c00-4f18-11e9-9701-a1789422a607.png)

Thanks for help!

**Comment by andyhahaha** [comment]
Sorry!
I found that it is caused by permute function.

**Comment by andyhahaha** [comment]
Hi, Zhihang
I do the permute by myself. It seems pretty good.
But after permute, pytorch need contiguous() before view().
contiguous raise this bug again.
WARNING: CANNOT FOUND blob at layer None, this may cause a NoneType Error. This may caused by the previous operation which produce the blob(tensor) is not implemented in nn_tools. You can issue this at https://github.com/hahnyuan/nn_tools/issues.

Caffe don't need contiguous function. How can I get trough it.
It's almost done for my model. Thanks for help!

**Comment by kongjibai** [comment]
> Caffe don't need contiguous function. How can I get trough it.
> It's almost done for my model. Thanks for help!

Hi, I met the similar situation, did you convert pytorch model to caffe model successfully?

**Comment by andyhahaha** [comment]
I get rid of continuous functions and use reshape instead of view.
View can only operate on continuous tensor so it needs continuous before it.
Reshape don't need continuous.
